Support returning plain values from methods

t0yv0 opened the pull request:
Support returning plain values from methods.

Implements Node, Python and Go support.

Remaining:

- [x] test receiving unknowns
- [x] acceptance tests written and passing locally for Node, Python, Go clients against a Go server
- [x] acceptance tests passing in CI
- [x] tickets filed for remaining languages
   - [x] https://github.com/pulumi/pulumi-yaml/issues/499
   - [x] https://github.com/pulumi/pulumi-java/issues/1193
   - [x] https://github.com/pulumi/pulumi-dotnet/issues/170 

Known limitations:

- this is technically a breaking change in case there is code out there that already uses methods that return Plain: true

- struct-wrapping limitation: the provider for the component resource needs to still wrap the plain-returning Method response with a 1-arg struct; by convention the field is named "res", and this is how it travels through the plumbing

- resources cannot return plain values yet

- the provider for the component resource cannot have unknown configuration, if it does, the methods will not be called
   - Per Luke https://github.com/pulumi/pulumi/issues/11520 this might not be supported/realizable yet

<!--- 
Thanks so much for your contribution! If this is your first time contributing, please ensure that you have read the [CONTRIBUTING](https://github.com/pulumi/pulumi/blob/master/CONTRIBUTING.md) documentation.
-->

# Description

<!--- Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. -->

Fixes https://github.com/pulumi/pulumi/issues/12709

## Checklist

- [ ] I have run `make tidy` to update any new dependencies
- [ ] I have run `make lint` to verify my code passes the lint check
  - [ ] I have formatted my code using `gofumpt`

<!--- Please provide details if the checkbox below is to be left unchecked. -->
- [ ] I have added tests that prove my fix is effective or that my feature works
<!--- 
User-facing changes require a CHANGELOG entry.
-->
- [ ] I have run `make changelog` and committed the `changelog/pending/<file>` documenting my change
<!--
If the change(s) in this PR is a modification of an existing call to the Pulumi Cloud,
then the service should honor older versions of the CLI where this change would not exist.
You must then bump the API version in /pkg/backend/httpstate/client/api.go, as well as add
it to the service.
-->
- [ ] Yes, there are changes in this PR that warrants bumping the Pulumi Cloud API version
  <!-- @Pulumi employees: If yes, you must submit corresponding changes in the service repo. -->

t0yv0 commented on pkg/codegen/testing/test/sdk_driver.go:
```diff
@@ -155,6 +155,11 @@ var PulumiPulumiSDKTests = []*SDKTest{
+	{
+		Directory:   "methods-return-plain-resource",
+		Description: "Test returning plain Resource objects from methods",
+		Skip:        codegen.NewStringSet("dotnet/any"),
```
Link TODO ticket here

t0yv0 commented:
Another note to self: need to implement EKS use case to dogwood this. EKS can configure a provider.

t0yv0 commented on sdk/go/pulumi/context.go:
```diff
@@ -557,6 +557,48 @@ func (ctx *Context) Call(tok string, args Input, output Output, self Resource, o
 }
 
+// Not to be used directly; exposed to support code-generated methods that return a plain (not Output-wrapped) Resource.
+func (ctx *Context) CallReturnPlaingResource(
```
Unfortunate to expose it in more user-facing API.

t0yv0 replied:
This is now tucked away in codegen utilities.

t0yv0 commented:
Adding integration tests seems helpful, although as-written it introduces a runtime dependency on AWS which is not great for this project; I'll refactor to use some more benign resources like those from the random provider; and cover the other languages.

t0yv0 commented on pkg/codegen/testing/test/testdata/methods-return-plain-resource/schema.json:
```diff
@@ -0,0 +1,44 @@
+        }
+      },
+      "outputs": {
+         "$ref": "/tls/v4.10.0/schema.json#/provider",
```
This is the example schema.

t0yv0 commented on tests/integration/construct_component_configure_provider/go/main.go:
```diff
@@ -0,0 +1,61 @@
+		key, err := tls.NewPrivateKey(ctx, "my-private-key", &tls.PrivateKeyArgs{
+			Algorithm:  pulumi.String("ECDSA"),
+			EcdsaCurve: pulumi.String("P384"),
+		}, pulumi.Provider(prov))
```
This is the intended use case, where Configurer remote component resource preconfigured a TLS provider, and the plain-returning method extracted a reference to it, which we can use to provision other resources.

t0yv0 commented:
Another flake on https://github.com/pulumi/pulumi/actions/runs/5755094465/job/15602064408?pr=13592#step:11:36

Frassle commented on pkg/codegen/schema/schema.go:
```diff
@@ -535,6 +535,8 @@ type Function struct {
 	// The return type of the function, if any.
 	ReturnType Type
+	// The return type is plain and not wrapped in an Output. This is only supported for resources.
+	ReturnTypePlain bool
```
This feels really sad. I guess all of codegen is written to assume that Resource types are always in Outputs so you can't mark the Type as plain?

t0yv0 replied:
Yeah. I would not say it's impossible to get there but it's certainly much less surgical. 

I'm less worried about editing codegen assumptions and more worried about editing per-language SDKs and serialization machinery that definitely assumes Resources have bags of outputs inside. I can't say I've explored this angle very deeply yet. Perhaps there's a way to get there in subsequent change.

Maybe there's a way to simply extend the code generation to add getters for them without interacting with the SDK machinery.

t0yv0 replied:
Like something along these lines could be a pure-codegen change:

```typescript
export class WorkforcePool extends pulumi.CustomResource {
    public readonly someResourceProperty!: pulumi.Output<Resource | undefined>;

    // add this
    function getSomeResourceProperty(): Promise<Resource> {..}
}
```

We add new class members. It's not entirely aesthetically satisfying. If we don't do it and edit the type of someResourceProperty then the generic machinery for setting it has to be likewise carefully edited.

Frassle replied:
Yeh I don't see any simple way to clean this up, something to try and come back to later I guess.

t0yv0 replied:
Ah perhaps I misunderstood if you're wondering whether we can support this for methods that return something like `Output<string>` to return a `Promise<string>`. I think his may be considered, this restriction can be lifted here. The current code as written can probably force `Output<string>` just as well to a `Promise<string>`.

As with any type the questions are what to do with unknown, secret bit, and dependencies.

We currently blame the provider for unknown and that is likely the right approach for every type.

For secret bits and dependencies, I was thinking that we can "inject" them into the Resource somehow in the future, if we only guarantee this functionality for resources. That being said I didn't find an easy way to do this.

Our other options are for secret bits and non-empty dependencies apply equally well to Resource or String etc.:

- fail
- warn
- warn very quietly (log at -v=9)

t0yv0 replied:
I'll think about this a bit and let other reviewers pipe in.

Frassle replied:
>We currently blame the provider for unknown and that is likely the right approach for every type.

Yeh that feels right. 
Was more just that we need this extra flag for plain resources, but plain strings are just encoded in ReturnType.

t0yv0 replied:
I've merged this, revising the code, this should support all values now not just resources.

Frassle commented on pkg/codegen/schema/schema.go:
```diff
@@ -535,6 +535,8 @@ type Function struct {
 	Outputs *ObjectType
 	// The return type of the function, if any.
 	ReturnType Type
+	// The return type is plain and not wrapped in an Output. This is only supported for resources.
```
FunctionSpec unmarshal code must need updating from this, otherwise this won't actually generate correctly for Go/Node SDKs.

t0yv0 replied:
Can you elaborate? The tests are verifying it working end to end.

bindFunctionDef sets ReturnTypePlain when producing *Function:

[[file:~/code/pulumi/pkg/codegen/schema/bind.go::func (t *types) bindFunctionDef(token string) (*Function, hcl.Diagnostics, error) {]]

In can see custom marshal code like unmarshalFunctionSpec but it deals with *FunctionSpec, not *Function. 

[[file:~/code/pulumi/pkg/codegen/schema/schema.go::func unmarshalFunctionSpec(funcSpec *FunctionSpec, data map\[string\]RawMessage) error {]]
https://github.com/pulumi/pulumi/blob/21770edc130518e2faa0c0506d60bb2af2997434/pkg/codegen/schema/schema.go#L1646-#L1646

Now *FunctionSpec doesn't have any changes.

t0yv0 replied:
I'm assuming Function is only produced from FunctionSpec via bindFunctionDef, then we should have all construction paths covered.

Frassle replied:
> Can you elaborate? The tests are verifying it working end to end.

The tests aren't really end to end at the moment, "gen-sdk" does a roundtrip via json marshaling to send schemas over grpc that you won't be seeing with the standard codegen tests.

Look at pkg/codegen/schema/schema.go around line 1176, you'll see you need to slot in pretty much the inverse of what you've done in bindFunctionDef.

t0yv0 replied:
Ok, I see. I'll look into it.

t0yv0 replied:
Thank you, edited pkg/codegen/schema/schema.go and threw in a test.

t0yv0 commented on pkg/codegen/python/gen.go:
```diff
@@ -1166,6 +1166,10 @@ func (mod *modContext) genResource(res *schema.Resource) (string, error) {
 			mod.collectImportsForResource(returnType.Properties, imports, false /*input*/, res)
+		} else if t, ok := method.Function.ReturnsPlainType(); ok {
+			if rt, ok := t.(*schema.ResourceType); ok {
+				imports.addResource(mod, rt)
```
Q: can this be an external type that needs imports as well, while not being a resource type?

t0yv0 commented:
Appreciate another pass. I've merged sub-PR and cleaned up the descriptions. I'll cross-link TODOs to tickets as the last step before merging after addressing all feedback. 

CC @Frassle

t0yv0 commented on pkg/codegen/nodejs/gen.go:
```diff
@@ -1022,12 +1035,21 @@ func (mod *modContext) genResource(w io.Writer, r *schema.Resource) (resourceFil
+
+		if doReturnPlainType {
+			prop := camel("res")
+			fmt.Fprintf(w, ", {plainResourceField: %q}", prop)
```
This shouldn't be called "plainResourceField" since the returned value can now be not a resource but any result really.

t0yv0 commented on pkg/codegen/go/gen.go:
```diff
@@ -2121,6 +1989,170 @@ func (pkg *pkgContext) genResource(w io.Writer, r *schema.Resource, generateReso
+	}
+
+	if doReturnPlainType {
+		fmt.Fprintf(w, "\tctx.CallReturnPlainResource(%q, %s, %s{}, r, reflect.ValueOf(&o), &e)\n",
```
CallReturnPlainResource is no longer a good name since we're returning any values not just a Resource. Also consider moving this out of SDK and into codegen utilities module.

t0yv0 commented on sdk/go/pulumi/context.go:
```diff
@@ -560,6 +560,48 @@ func (ctx *Context) Call(tok string, args Input, output Output, self Resource, o
+		// Ignoring deps; would be better to attach them to the *T resource.
+		// Ignoring secret; perhaps would be better to mark all outputs in the *T resource as secret.
+
+		contract.Assertf(known, "Plain resource method %q incorrectly returned an unknown Resource value. "+
```
s/unknown Resource value/unknown value/

t0yv0 commented on pkg/codegen/python/gen.go:
```diff
@@ -1684,7 +1712,10 @@ func (mod *modContext) genMethods(w io.Writer, res *schema.Resource) {
 
-		if returnType == nil {
+		if doReturnPlainType {
+			fmt.Fprintf(w, "        return pulumi.runtime.call('%s', __args__, res=__self__%s, plainResourceField='%s')\n",
```
plainResourceField again --> plainField or something else

t0yv0 commented:
Comment from @Zaid-Ajaj 👍 
```
func (c *Configurer) TlsProvider(ctx *pulumi.Context, args *TlsProviderArgs) (*TlsProviderResult, error) {
	return &TlsProviderResult{
		Result: c.TlsProviderOutput,
	}, nil
}
can't this be the following?
func (c *Configurer) TlsProvider(ctx *pulumi.Context, args *TlsProviderArgs) (*TlsProviderOutput, error) {
	return c.TlsProviderOutput, nil
}
```

We do need to wrap x to `{"res": x}` for the transport layer, but there's an opportunity to simplify the interface here perhaps if return pulumiprovider.NewCallResult(result) automatically did the wrapping.

t0yv0 commented:
@Frassle @Zaid-Ajaj I've simplified this a ton and removed the edits from the SDKs so that the change is atomic over codegen; tests still pass.

Now looking at generalizing to object output types with multiple properties. It seems like it should work similarly with little effort, just having some hiccups with the schema parser, when you specify "properties" it doesn't like `"plain": true` anymore, nor does it like `"plain": ["*"]`, triggers:

```
Received unexpected error:
                                <nil>: #/functions/metaprovider:index:Configurer%2FobjectMix/outputs/plain: plain has been removed; the property type must be marked as plain instead;  
```


```
    "metaprovider:index:Configurer/objectMix": {
      "inputs": {
        "properties": {
          "__self__": {
            "$ref": "#/resources/metaprovider:index:Configurer"
          }
        }
      },
      "outputs": {
        "type": "object",
        "plain": ["*"],
        "properties": {
          "provider": {
            "$ref": "/tls/v4.10.0/schema.json#/provider"
          },
          "meaningOfLife": {
            "type": "integer"
          }
        }
      }
    }
```

Need some minor surgery to allow the user to specify they want the output plain.

t0yv0 commented:
Ready for another review round. This now supports both singleton and object-returning plain method calls. And wrapping in magic "res" field is simplified for provider authors with another helper.

t0yv0 commented:
@Frassle @Zaid-Ajaj appreciate a review here.

Zaid-Ajaj reviewed (approved):
Looks ready to go ✅ still need to open issues for follow-up tasks but that is a non-blocker

t0yv0 commented:
FunctionSpec was not roundtripping through JSON serialization over new ObjetTypeSpecIsPlain form. This is now fixed up.

t0yv0 commented on pkg/codegen/schema/schema_test.go:
```diff
@@ -996,3 +996,280 @@ func TestBindDefaultInt(t *testing.T) {
 	}
 }
+
+func TestFunctionSpecToJSONAndYAMLTurnaround(t *testing.T) {
```
Allright this turns out to be particularly tricky since json and yaml packages behave differently around the edges; hopefully this test matrix covers all the interesting variations.

Frassle commented on changelog/pending/20230803--sdkgen-go-nodejs-python--resource-methods-can-return-plain-values.yaml:
```diff
@@ -0,0 +1,4 @@
+changes:
+- type: feat
+  scope: sdkgen/go,nodejs,python
+  description: "Resource methods with plain: true outputs can now return plain values without an Output wrapper. In particular, this feature enables resoruce methods to serve as explicit provider factories by returning preconfigured explicit providers."
```
```suggestion
  description: "Resource methods with plain: true outputs can now return plain values without an Output wrapper. In particular, this feature enables resource methods to serve as explicit provider factories by returning preconfigured explicit providers."
```

t0yv0 commented:
I've started rebasing this PR on master, but it looks like it has some meaningful conflicts now with Go generics. For starters I've re-inlined genMethod to reduce the diff. I need a bit more time to double-check it again and possibly add a test case for generics support or else defer this to a ticket. It might not work properly with generics.

t0yv0 commented:
I have performed enhanced downstream checks with this. Apart from one failure it's all green.

Request timed out:
https://github.com/pulumi/pulumi-f5bigip/actions/runs/6905614344/job/18789013409#step:23:227

I think the failure is not related.
